module-services/service-desktop/WorkerDesktop.cpp

```diff
@@ -113,95 +115,126 @@ void WorkerDesktop::reset()
 
 bool WorkerDesktop::handleMessage(uint32_t queueID)
 {
+    auto result       = false;
     auto &queue       = queues[queueID];
     const auto &qname = queue->GetQueueName();
 
     if (qname == sdesktop::RECEIVE_QUEUE_BUFFER_NAME) {
-        std::string *receivedMsg = nullptr;
-        if (!queue->Dequeue(&receivedMsg, 0)) {
-            LOG_ERROR("handleMessage failed to receive from \"%s\"", sdesktop::RECEIVE_QUEUE_BUFFER_NAME);
-            return false;
-        }
+        result = handleReceiveQueueMessage(queue);
+    }
+    else if (qname == sdesktop::SEND_QUEUE_BUFFER_NAME) {
+        result = handleSendQueueMessage(queue);
+    }
+    else if (qname == SERVICE_QUEUE_NAME) {
+        result = handleServiceQueueMessage(queue);
+    }
+    else if (qname == sdesktop::IRQ_QUEUE_BUFFER_NAME) {
+        result = handleIrqQueueMessage(queue);
+    }
+    else {
+        LOG_INFO("handeMessage got message on an unhandled queue");
+    }
 
-        using namespace sdesktop::endpoints;
-        auto factory = EndpointFactory::create(securityModel.getEndpointSecurity());
-        auto handler = std::make_unique<MessageHandler>(ownerService, messageProcessedCallback, std::move(factory));
+    return result;
+}
 
-        parser.setMessageHandler(std::move(handler));
-        parser.processMessage(std::move(*receivedMsg));
+bool WorkerDesktop::handleReceiveQueueMessage(std::shared_ptr<sys::WorkerQueue> &queue)
+{
+    if (not initialized) {
+        return false;
+    }
+    std::string *receivedMsg = nullptr;
+    if (!queue->Dequeue(&receivedMsg, 0)) {
+        LOG_ERROR("handleMessage failed to receive from \"%s\"", sdesktop::RECEIVE_QUEUE_BUFFER_NAME);
+        return false;
+    }
 
-        delete receivedMsg;
+    using namespace sdesktop::endpoints;
+    auto factory = EndpointFactory::create(securityModel.getEndpointSecurity());
+    auto handler = std::make_unique<MessageHandler>(ownerService, messageProcessedCallback, std::move(factory));
+
+    parser.setMessageHandler(std::move(handler));
+    parser.processMessage(std::move(*receivedMsg));
+
+    delete receivedMsg;
+    return true;
+}
+bool WorkerDesktop::handleSendQueueMessage(std::shared_ptr<sys::WorkerQueue> &queue)
+{
+    if (not initialized) {
+        return false;
+    }
+    std::string *sendMsg = nullptr;
+    if (!queue->Dequeue(&sendMsg, 0)) {
+        LOG_ERROR("handleMessage xQueueReceive failed for %s.", sdesktop::SEND_QUEUE_BUFFER_NAME);
+        return false;
     }
-    else if (qname == sdesktop::SEND_QUEUE_BUFFER_NAME) {
-        std::string *sendMsg = nullptr;
-        if (!queue->Dequeue(&sendMsg, 0)) {
-            LOG_ERROR("handleMessage xQueueReceive failed for %s.", sdesktop::SEND_QUEUE_BUFFER_NAME);
-            return false;
-        }
 
-        bsp::usbCDCSend(sendMsg);
-        delete sendMsg;
+    bsp::usbCDCSend(sendMsg);
+    delete sendMsg;
```

Comment:
Here is the same story, however here the message seems to be copied into `usbCDCSend`. Maybe it should be moved as well?

Reply:
Done

Reply:
Actually `bsp::usbCDCSend` probably takes it by `const&` so this change is probably only semantic.

Reply:
Actually actually no, it seems that `bsp::usbCDCSend` expects a pointer to `std::string`. So it's not a matter of ownership but a matter of consistency of handling messages. Above it's taken by reference but here by pointer.